### PR TITLE
Support quotes in Cumulocity failureReason field

### PR DESCRIPTION
**Pull request by jarhodes314**
## Proposed changes
This PR simplifies the CSV serialization logic for `SmartRestSetOperationToFailed`, allowing quotes inside the `failure_reason` field to be preserved.

It also makes a small tweak to an error message modified as part of #2474 to make it more generic, referring to `config-manager` instead of `tedge-configuration-plugin` as suggested by @albinsuresh. 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#2487 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by rina23q — APPROVED** [review]
The only concern was to address `,` inside failure reason, however, I see that this unit test proves that the `csv` crates adds double quotes if a string contains a comma. So I would approve the PR.

```rust
    #[test]
    fn serialize_smartrest_set_operation_to_failed_with_comma_reason() {
        let smartrest = SmartRestSetOperationToFailed::new(
            CumulocitySupportedOperations::C8ySoftwareUpdate,
            "Failed to install collectd, modbus, and golang.".into(),
        )
        .to_smartrest()
        .unwrap();
        assert_eq!(
            smartrest,
            "502,c8y_SoftwareUpdate,\"Failed to install collectd, modbus, and golang.\"\n"
        );
    }
```

File: crates/core/c8y_api/src/smartrest/smartrest_serializer.rs [hunk 1]
```diff
@@ -1,15 +1,76 @@
...
 
 pub type SmartRest = String;
 
-#[derive(Debug)]
+pub fn request_pending_operations() -> &'static str {
+    "500"
+}
+
+/// Generates a SmartREST message to set the provided operation to executing
+pub fn set_operation_executing(operation: impl C8yOperation) -> String {
```
Comment (jarhodes314) [inline crates/core/c8y_api/src/smartrest/smartrest_serializer.rs:18]: The main additional changes from the last review are things like this, which replace the Serialization structs with functions.

File: crates/core/c8y_api/src/smartrest/smartrest_serializer.rs [hunk 2]
```diff
@@ -32,185 +93,124 @@ impl From<CumulocitySupportedOperations> for &'static str {
...
-            message_id: "502",
-            operation: operation.into(),
-            reason,
+impl Serialize for EmbeddedCsv {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        S::Error: serde::ser::Error,
+    {
+        let record = dbg!(parse_single_record(&self.0)?);
```
Comment (jarhodes314) [inline crates/core/c8y_api/src/smartrest/smartrest_serializer.rs:161]: Oops, didn't mean to leave that `dbg!` in. Just leaving this note here so I remember to get rid of it.

File: crates/core/c8y_api/src/smartrest/smartrest_serializer.rs [hunk 3]
```diff
@@ -32,185 +93,124 @@ impl From<CumulocitySupportedOperations> for &'static str {
...
-        }
+impl From<String> for TextOrCsv {
+    fn from(value: String) -> Self {
+        Self::Text(value)
     }
 }
 
-impl<'a> SmartRestSerializer<'a> for SmartRestSetOperationToSuccessful {}
+#[derive(Debug, Eq, PartialEq)]
+pub struct EmbeddedCsv(String);
```
Comment (jarhodes314) [inline crates/core/c8y_api/src/smartrest/smartrest_serializer.rs:147]: Additionally I've added this `EmbeddedCsv` struct for handling embedded CSVs (which come from custom operations with an result type of CSV. This will deserialize and reserialize the CSV when we add the fields to ensure a consistent format, and to ensure we don't generate a broken message (or multiple messages if the output contains newlines).

File: crates/core/c8y_api/src/smartrest/smartrest_serializer.rs [hunk 4]
```diff
@@ -219,76 +219,122 @@ mod tests {
 
     #[test]
```
Comment (jarhodes314) [inline crates/core/c8y_api/src/smartrest/smartrest_serializer.rs:220]: These tests should give a broad overview of the updated functionality. Mostly, it means we no longer have unnecessary quotes and newlines, and generating the messages is simpler.

File: crates/extensions/c8y_mapper_ext/src/service_monitor.rs [hunk 5]
```diff
@@ -68,55 +68,55 @@ mod tests {
         r#"{"pid":"1234","status":"up"}"#,
         "c8y/s/us/test_device:device:main:service:tedge-mapper-c8y",
         r#"104,up"#;
-        "service-monitoring-thin-edge-device"
+        "service monitoring thin-edge device"
```
Comment (jarhodes314) [inline crates/extensions/c8y_mapper_ext/src/service_monitor.rs:71]: Not related, I just failed to spot (reading the code) that this was a test name and not just more data when it was kebab-case.

File: crates/extensions/tedge_mqtt_ext/src/test_helpers.rs [hunk 6]
```diff
@@ -47,9 +47,7 @@ pub fn assert_message_contains_str(message: &Message, expected: (&str, &str)) {
     let payload = message.payload_str().expect("non UTF-8 payload");
     assert!(
         payload.contains(expected_payload),
-        "Payload assertion failed.\n Actual: {} \n Expected: {}",
-        payload,
-        expected_payload
+        "Payload assertion failed.\n Actual: {payload:?} \n Expected: {expected_payload:?}",
```
Comment (jarhodes314) [inline crates/extensions/tedge_mqtt_ext/src/test_helpers.rs:50]: I added debug printing here because the assertion message was very unclear when it failed due to a newline in the expected payload.

**Review by jarhodes314 — COMMENTED** [review]
I've added a few comments to guide you through the changes since this was last reviewed. Most of the changes are very monotonous, so the code is lengthy, but they should all stem from the changes in `smartrest_serializer.rs`, and I've highlighted the specific changes there.

File: crates/core/c8y_api/src/smartrest/message.rs [hunk 7]
```diff
@@ -167,63 +166,62 @@ mod tests {
...
-        let input: Vec<u8> = (0x00..0xff).collect();
-        let sanitized = sanitize_for_smartrest(input, MAX_PAYLOAD_LIMIT_IN_BYTES);
-        let re = Regex::new(r"[^\x20-\x7E\xA0-\xFF\t\n\r]").unwrap();
-        assert!(!re.is_match(&sanitized));
+        let input: Vec<u8> = (0x00..=0x7f).collect();
+        let sanitized = sanitize_bytes_for_smartrest(&input, MAX_PAYLOAD_LIMIT_IN_BYTES);
+        assert_eq!(
+            sanitized,
+            "\t\n\r !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
+        );
```
Comment (jarhodes314) [inline crates/core/c8y_api/src/smartrest/message.rs:181]: I modified the range this test covers and the assertion because the test was originally supplying invalid UTF-8 (so the sanitize function was returning an empty string), and I don't think this was the intention. The Regex matching obfuscated this since the empty string trivially passed the assertion, so I expanded this into a string literal as this guarantees the output is what we want.

**Review by albinsuresh — APPROVED** [review]
LGTM

File: crates/core/c8y_api/src/smartrest/smartrest_serializer.rs [hunk 8]
```diff
@@ -204,90 +219,122 @@ mod tests {
...
+        assert_eq!(smartrest, "503,c8y_RelayArray,true,false,true");
+    }
+
+    #[test]
+    fn serialize_smartrest_set_custom_operation_to_successful_with_multi_record_csv_payload() {
+        succeed_operation(
+            "c8y_RelayArray",
+            TextOrCsv::Csv(EmbeddedCsv("true\n1,2,3".to_owned())),
+        )
+        .unwrap_err();
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/core/c8y_api/src/smartrest/smartrest_serializer.rs:280]: I agree that assert is more explicit about what we're doing, but if the assertion fails, the only information you get is that the assertion failed. By comparison, unwrap_err will debug print the Ok value that we tried to extract the error from, which is likely useful.

File: crates/core/c8y_api/src/smartrest/smartrest_serializer.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/c8y_api/src/smartrest/smartrest_serializer.rs:18]: Indeed, the intent is better conveyed compared to these SmartRest codes such as 501.

**Review by didier-wenzek — APPROVED** [review]
